### PR TITLE
Add math operator patches

### PR DESCRIPTION
**Pull request by reyoung**
Users can use `a+b`, `a*10`.

### CODE REVIEW

**Comment by tonyyang-svail** [comment]
will this patch handle `b = -a`?

File: python/paddle/v2/fluid/layers/math_op_patch.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+from ..framework import Variable, unique_name
+from ..registry import OpProtoHolder
+
+__all__ = ['monkey_patch_variable']
+
+
+def monkey_patch_variable():
+    def new_name():
+        return unique_name("tmp")
```
Comment (JiayiFeng) [inline python/paddle/v2/fluid/layers/math_op_patch.py]: Any more meaningful name?

File: python/paddle/v2/fluid/layers/math_op_patch.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+                   "shape": [1],
+                   "dtype": dtype})
+        return var
+
+    def create_tensor(block, value, dtype, shape):
+        value = float(value)
+        tmp_name = new_name()
+        var = block.create_var(name=tmp_name, shape=shape, dtype=dtype)
+        block.append_op(
+            type="fill_constant",
```
Comment (JiayiFeng) [inline python/paddle/v2/fluid/layers/math_op_patch.py]: Why do we use `fill_constant_op` in `create_tensor` while `fill_op` in `create_scalar`?

File: python/paddle/v2/fluid/layers/math_op_patch.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            type='fill_constant_batch_size_like',
+            outputs={'Out': [var]},
+            inputs={'Input': [ref_var]},
+            attrs={'shape': ref_var.shape,
+                   'value': value})
+        return var
+
+    def astype(self, dtype):
+        """
+        Cast a variable to data type.
```
Comment (JiayiFeng) [inline python/paddle/v2/fluid/layers/math_op_patch.py]: Cast a variable to **a specified** data type.

File: python/paddle/v2/fluid/layers/math_op_patch.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                            dtype=lhs_dtype,
+                            shape=self.shape)
+                    else:
+                        other_var = create_tensor_with_batchsize(
+                            self, other_var, lhs_dtype)
+                else:
+                    # add fill_op to self.block
+                    other_var = create_scalar(
+                        self.block, value=other_var, dtype=lhs_dtype)
+
```
Comment (JiayiFeng) [inline python/paddle/v2/fluid/layers/math_op_patch.py]: Maybe we shall add a type check here to make sure the rhs have been cast to Variable correctly. For some users may send `np.array` to the `other_var`.

File: python/paddle/v2/fluid/layers/math_op_patch.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+                    if not has_batch_size:
+                        other_var = create_tensor(
+                            self.block,
+                            other_var,
+                            dtype=lhs_dtype,
+                            shape=self.shape)
+                    else:
+                        other_var = create_tensor_with_batchsize(
+                            self, other_var, lhs_dtype)
+                else:
```
Comment (JiayiFeng) [inline python/paddle/v2/fluid/layers/math_op_patch.py]: Why does the `other_var` is cast to `Variable` by different methods in the two branches of `if reverse:`?

**Comment by reyoung** [comment]
> will this pr handle b = -a?

No, it won't. But following this PR, we can overload all operators in Python. We can give an API just like numpy.

File: python/paddle/v2/fluid/layers/math_op_patch.py (hunk 5 above)
Comment (reyoung) [reply python/paddle/v2/fluid/layers/math_op_patch.py]: The reversing operator in Python means the left-hand operand is a Python value and the right-hand operand is a Variable.

In the elementwise-operators, the right-hand operand can be broadcasted. Supposed we write an `a - 10`, every element in `a` will be minus 10.  However, for `10 - a`, we do not implement the broadcast in C++ operators. So we should explicitly broadcast `10` to the same shape of `a`.

Getting the shape of a variable has two situations. 

1. The variable is a parameter. Its shape is decided at compile time and not related to the batch size. So we use `fill_constant`.
2. The variable is a layer output. Its shape is decided at runtime and related to the batch size. So we use `fill_constant_batch_size_like`.

File: python/paddle/v2/fluid/layers/math_op_patch.py (hunk 4 above)
Comment (reyoung) [reply python/paddle/v2/fluid/layers/math_op_patch.py]: The `create_scalar` and `create_tensor` have checked the value can be cast to `float`.

File: python/paddle/v2/fluid/layers/math_op_patch.py (hunk 2 above)
Comment (reyoung) [reply python/paddle/v2/fluid/layers/math_op_patch.py]: Cool

File: python/paddle/v2/fluid/layers/math_op_patch.py (hunk 3 above)
Comment (reyoung) [reply python/paddle/v2/fluid/layers/math_op_patch.py]: Cool, thanks.

File: python/paddle/v2/fluid/layers/math_op_patch.py (hunk 1 above)
Comment (reyoung) [reply python/paddle/v2/fluid/layers/math_op_patch.py]: Done.

File: python/paddle/v2/fluid/layers/math_op_patch.py (hunk 5 above)
Comment (JiayiFeng) [reply python/paddle/v2/fluid/layers/math_op_patch.py]: Great idea. Impressive.

File: python/paddle/v2/fluid/layers/math_op_patch.py (hunk 4 above)
Comment (JiayiFeng) [reply python/paddle/v2/fluid/layers/math_op_patch.py]: OK, I see.

**Pull request merged**